fdbserver/MutationTracking.cpp

```diff
@@ -0,0 +1,100 @@
+		return std::move(TraceEvent("MutationTracking").detail("At", context).detail("Version", version).detail("MutationType", typeString[mutation.type]).detail("KeyBegin", mutation.param1).detail("KeyEnd", mutation.param2));
+	} else if (mutation.param1 == debugKey || mutation.param1 == debugKey2) {
+		return std::move(TraceEvent("MutationTracking").detail("At", context).detail("Version", version).detail("MutationType", typeString[mutation.type]).detail("Key", mutation.param1).detail("Value", mutation.param2));
+	} else {
+		return std::move(TraceEvent());
+	}
+}
+
+TraceEvent debugKeyRangeEnabled( const char* context, Version version, KeyRangeRef const& keys ) {
+	if (keys.contains(debugKey) || keys.contains(debugKey2)) {
+		return std::move(debugMutation(context, version, MutationRef(MutationRef::DebugKeyRange, keys.begin, keys.end) ));
+	} else {
+		return std::move(TraceEvent());
+	}
+}
+
+TraceEvent debugMessagesAndTagsEnabled( const char* context, Version version, StringRef commitBlob ) {
+	BinaryReader rdr(commitBlob, AssumeVersion(currentProtocolVersion));
+	while (!rdr.empty()) {
+		if (*(int32_t*)rdr.peekBytes(4) == VERSION_HEADER) {
+			int32_t dummy;
+			rdr >> dummy >> version;
+			continue;
+		}
+		TagsAndMessage msg;
+		msg.loadFromArena(&rdr, nullptr);
+		bool logAdapterMessage = std::any_of(
+				msg.tags.begin(), msg.tags.end(), [](const Tag& t) { return t == txsTag || t.locality == tagLocalityTxs; });
+		StringRef mutationData = msg.getMessageWithoutTags();
+		uint8_t mutationType = *mutationData.begin();
+		if (logAdapterMessage) {
+			// Skip the message, as there will always be an idential non-logAdapterMessage mutation
+			// that we can match against in the same commit.
+		} else if (LogProtocolMessage::startsLogProtocolMessage(mutationType)) {
+			BinaryReader br(mutationData, AssumeVersion(rdr.protocolVersion()));
+			LogProtocolMessage lpm;
+			br >> lpm;
+			rdr.setProtocolVersion(br.protocolVersion());
+		} else {
+			MutationRef m;
+			BinaryReader br(mutationData, AssumeVersion(rdr.protocolVersion()));
+			br >> m;
+			TraceEvent&& event = debugMutation(context, version, m);
+			if (event.isEnabled()) {
+				return std::move(event.detail("MessageTags", msg.tags));
+			}
+		}
+	}
+	return std::move(TraceEvent());
+}
+
+#if MUTATION_TRACKING_ENABLED
+TraceEvent debugMutation( const char* context, Version version, MutationRef const& mutation ) {
+	return debugMutationEnabled( context, version, mutation );
+}
+TraceEvent debugKeyRange( const char* context, Version version, KeyRangeRef const& keys ) {
+	return debugKeyRangeEnabled( context, version, keys );
+}
+TraceEvent debugMessagesAndTags( const char* context, Version version, StringRef commitBlob ) {
+	return debugMessagesAndTagsEnabled( context, version, commitBlob );
+}
+#else
+TraceEvent debugMutation( const char* context, Version version, MutationRef const& mutation ) { return std::move(TraceEvent()); }
```

Comment:
Why the explicit std::move, is there doubt that RVO will occur?

Reply:
It's at the mercy of compliers.

Reply:
I could be wrong but I thought RVO would handle things like `TraceEvent e = debugkeyRange(..)` in this scenario.

Reply:
If it doesn't do the optimization, are you going to get an extra trace event logged due to the destruction?

Reply:
I think so.

Reply:
@dongxinEric Sure, but if we are assuming the worst then I guess a lot of other code should probably be more explicit about such things as well.

In any case, I'm not sure that this pattern is what we want here.  Moving a TraceEvent sounds better than copying one, but even constructing one is not exactly trivial and moving one [isn't either](https://github.com/apple/foundationdb/blob/master/flow/Trace.cpp#L764).

Since disabled TraceEvents do not log details, perhaps when mutation tracking is disabled these methods could return a reference to a static TraceEvent that exists to absorb/ignore detail calls, rather than initializing a new one only to have some ignored details added before it is destroyed.  

Or perhaps even better would be for these methods to return a type that looks like a TraceEvent but does nothing, so even the detail() argument initialization should be optimized out.  If the callers used `auto e = fun(...)` rather than `TraceEvent e = fun(...)` this would work, though it is at least a bit awkward.

However, these suggestions still have an overhead problem where statements like `e.detail("Foo", x.toString())` (which I see in this changeset) will still execute the `.toString()` call (unless compiler and link time optimization is being amazing) even though the detail won't be added.

I should point out that my assumption here is that since every key that passes through every role of fdb is going to have `debugMutation()` called on it several times, constructing a default TraceEvent and making some `detail()` calls each time that happens is not a trivial amount of overhead.

Reply:
Since the return type of the method is `TraceEvent` and not `TraceEvent &&`, I'm not even sure the `std::move()` is doing anything.

Reply:
As of C++17, the RVO in this case is mandatory: https://en.cppreference.com/w/cpp/language/copy_elision

We could even go so far as deleting the copy/move constructors, and these functions would still work.

Reply:
> Why the explicit std::move, is there doubt that RVO will occur?

Because if it's not there, then gcc threw an error that the copy constructor was private/deleted.

Reply:
If we really want the overhead of this to go away, I can instead

```
#define UNHYGENIC_DEBUG_MUTATION(context, version, mutationref) !MUTATION_TRACKING_ENABLED ? (void)TraceEvent() : debugMutationEnabled(context, version, mutationref)

UNHYGENIC_DEBUG_MUTATION("ProxyCommit", version, m).detail("Foo", foo);
```

But I was hoping it wouldn't come to that.  I'll see if I can play more with c++17 and RVO to see if I can massage the code into something that gcc will accept and doesn't require a move or copy.

My overall preference is to keep the code as something that's compiled and then dead code eliminated, so that people are forced to maintain it as .e.g variable names change.

Reply:
I do think we need to eliminate this overhead, I would be very surprised if this much work, many times over, per KV pair, is trivial.

I absolutely agree with the goal of having code that is always compiled but that is optimized out when not needed.

My understanding is that the compiler would have no way of knowing that the `.detail()` calls on the default-constructed TraceEvent have no side effects and can safely be removed, since the `detail()` methods are not defined in `Trace.h`, so I don't think it can just decide to not call them.

There is of course always the option of a run-time enabled check using `if` rather than `#if`, such as
```
if(MUTATION_TRACKING_ENABLED) {
...
}
```
Since `MUTATION_TRACKING_ENABLED` is a constant the code block should be compiled but optimized out.

Here's another option that looks cleaner and I think would work.  First, MUTATION_TRACKING_ENABLED has to be defined to 0 if it is not already defined.  Then make `TraceEvent` cast to `bool` defined as
```
operator bool() const {
  return enabled;
}
```
which I think makes sense in a bool context.  Then, we can do this:
```
MUTATION_TRACKING_ENABLED && debugMutation(...).detail(...).detail(..);
```
and the right hand side should be compiled and then optimized out.

Reply:
I ended up doing a hybrid of the two.  I added a macro that does `MUTATION_TRACKING_ENABLED && debugMutation(...)`, so writing `DEBUG_MUTATION(x,y,z).detail("foo", "bar")` expands to `ENABLED && debugMutation(x,y,z).debug("foo", "bar")`.  This breaks being able to do `TraceEvent&& e = DEBUG_MUTATION(x,y,z)`, so one has to manually `if (MUTATION_TRACKING_ENABLED)`, and then call `debugMutation` instead.  One shouldn't need to do this often, so I think it's okay.